### PR TITLE
TypeInfo.GetAncestors() -> TypeInfo.GetAncestor(int index)

### LINKED ISSUE #2152: Improve GetAncestors
Look at `GetAncestors()` usages. Only two places with index access. Let's refactor `GetAncestors` into `GetAncestor(int index)`. Also you can create `stackalloc` array instead of calling `ToArray` and then call `CopyTo`, so no array allocation 😄

_Originally posted by @sungam3r in https://github.com/graphql-dotnet/graphql-dotnet/pull/2139#discussion_r553385784_

### PR DESCRIPTION
**Pull request by sungam3r**
fixes #2152

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.16299.726 (1709/FallCreatorsUpdate/Redstone3)
Intel Core i7 CPU 920 2.67GHz (Nehalem), 1 CPU, 8 logical and 4 physical cores
Frequency=2607421 Hz, Resolution=383.5207 ns, Timer=TSC
.NET Core SDK=5.0.101
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  DefaultJob : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
```

```
before

|        Method |      Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |----------:|---------:|---------:|-------:|------:|------:|----------:|
| Introspection | 553.22 μs | 1.922 μs | 1.798 μs | 7.8125 |     - |     - |  34.13 KB |
|          Hero |  46.22 μs | 0.053 μs | 0.047 μs | 3.7231 |     - |     - |   15.3 KB |

after

|        Method |      Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |----------:|---------:|---------:|-------:|------:|------:|----------:|
| Introspection | 573.30 μs | 0.944 μs | 0.837 μs | 7.8125 |     - |     - |  33.44 KB |
|          Hero |  45.16 μs | 0.036 μs | 0.030 μs | 3.7231 |     - |     - |   15.3 KB |

after remove try/finally

|        Method |      Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |----------:|---------:|---------:|-------:|------:|------:|----------:|
| Introspection | 553.22 μs | 0.710 μs | 0.664 μs | 7.8125 |     - |     - |  33.44 KB |
|          Hero |  46.38 μs | 0.099 μs | 0.087 μs | 3.7231 |     - |     - |   15.3 KB |
```

You can try it yourself. Execution time is confusing again.

### CODE REVIEW

File: src/GraphQL/Validation/TypeInfo.cs [hunk 1]
```diff
@@ -29,12 +28,28 @@ public TypeInfo(ISchema schema)
         }
 
         /// <summary>
-        /// Returns a list of ancestors of the current node.
+        /// Returns an ancestor of the current node.
         /// </summary>
-        /// <returns></returns>
-        public INode[] GetAncestors()
+        /// <param name="index">Index of the ancestor; 0 for the node itself, 1 for the direct ancestor and so on.</param>
+        public INode GetAncestor(int index)
```
Comment (sungam3r) [inline src/GraphQL/Validation/TypeInfo.cs:34]: @Shane32 `stackalloc` only works for primitive types and structures. But everything turned out to be even easier.

**Review by Shane32 — APPROVED** [review]

**Comment by Shane32** [comment]
I will try it later.  If it's working slower than LINQ, I'd be quite surprised.  I think you have the fastest implementation right now.  But it is possible that Stack.ToArray is faster than using an enumerator.  And if it is, it's probably due to less jumps and conditionals within the compiled assembly language instructions, so it doesn't have to empty the CPU pipeline as often.  I'll try it.

**Comment by Shane32** [comment]
I ran benchmarks:
- `tune` branch
- the develop branch as it appears in the `tune` branch history
- the develop branch as it appears in the `tune` branch history, along with the changes to GetAncestors that were in #2139 (that are now in the current develop branch)

```
tune

|        Method |      Mean |    Error |   StdDev |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|-------------- |----------:|---------:|---------:|-------:|-------:|------:|----------:|
| Introspection | 237.20 us | 0.540 us | 0.505 us | 5.3711 |      - |     - |  33.44 KB |
|     Fragments | 102.13 us | 0.243 us | 0.215 us | 3.6621 | 0.1221 |     - |   22.8 KB |
|          Hero |  19.98 us | 0.242 us | 0.227 us | 2.4719 | 0.0610 |     - |   15.3 KB |


develop (pre-validation_3)

|        Method |      Mean |    Error |   StdDev |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|-------------- |----------:|---------:|---------:|-------:|-------:|------:|----------:|
| Introspection | 239.25 us | 0.742 us | 0.620 us | 5.3711 |      - |     - |  34.13 KB |
|          Hero |  19.92 us | 0.205 us | 0.172 us | 2.4719 | 0.0610 |     - |   15.3 KB |


develop (pre-validation_3 with revised getAncestors from validation_3)


|        Method |      Mean |    Error |   StdDev |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|-------------- |----------:|---------:|---------:|-------:|-------:|------:|----------:|
| Introspection | 232.62 us | 0.698 us | 0.653 us | 5.3711 |      - |     - |   33.7 KB |
|          Hero |  19.54 us | 0.070 us | 0.066 us | 2.4719 | 0.0610 |     - |   15.3 KB |
```

I'm fine with whatever changes you want to make to this method, btw. You decide.  If you want me to run more benchmarks, let me know.

**Comment by sungam3r** [comment]
I think we should accept these changes. The more complex (nested) query, the greater the gain due to the fact that arrays are not copied. The most recent elements of the stack are always used.

**Comment by Shane32** [comment]
I also though of this code:
```cs
        public void GetAncestors(out INode parent, out INode grandparent)
        {
            var ancestorArray = _ancestorStack.ToArray();
            parent = ancestorArray[1];
            grandparent = ancestorArray[2];
        }
```
But this tested slower:
```
|        Method |      Mean |    Error |   StdDev |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|-------------- |----------:|---------:|---------:|-------:|-------:|------:|----------:|
| Introspection | 238.93 us | 1.084 us | 1.014 us | 5.3711 |      - |     - |  33.58 KB |
|          Hero |  19.18 us | 0.098 us | 0.092 us | 2.4719 | 0.0610 |     - |   15.3 KB |
```
Which makes no sense.  Maybe it's that ref/out parameters are slow. I dunno

**Comment by sungam3r** [comment]
I believe that a slight drop in performance of the most primitive queries can be allowed in favor of general normalization of work on an arbitrary query size.

**Comment by Shane32** [comment]
I'd also like to comment that in the one place it's used, you are allocating an enumeration object twice just to read the first and second entries.

**Comment by sungam3r** [comment]
Did you test my changes with try/finally removed ?

**Comment by sungam3r** [comment]
Stack enumerator is struct.

**Comment by Shane32** [comment]
> Did you test my changes with try/finally removed ?

Yes I just ran them - it had this history at the time I ran it:

image

**Comment by Shane32** [comment]
> Stack enumerator is struct.

Oh well that's cool! 👍

**Comment by Shane32** [comment]
I tested this also:
```cs
        public (INode parent, INode grandparent) GetAncestors()
        {
            var ancestorArray = _ancestorStack.ToArray();
            return (ancestorArray[1], ancestorArray[2]);
        }
```

Still repeatably tests slower than this code:
```cs
        public INode[] GetAncestors()
        {
            var ancestorArray = _ancestorStack.ToArray();
            var c = ancestorArray.Length;
            var ret = new INode[--c];
            for (int i = c; i > 0; --i)
            {
                ret[c - i] = ancestorArray[i];
            }
            return ret;
            // aka: return _ancestorStack.Skip(1).Reverse().ToArray();
        }
```

Weird.

Yeah merge whatever you want 👍

**Comment by Shane32** [comment]
I can run post-merge benchmarks now if you like.

**Comment by sungam3r** [comment]
OK

**Comment by Shane32** [comment]
```
tune

|        Method |      Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |----------:|---------:|---------:|-------:|------:|------:|----------:|
| Introspection | 180.17 us | 0.371 us | 0.347 us | 3.1738 |     - |     - |  19.59 KB |
|     Fragments |  80.66 us | 0.378 us | 0.354 us | 1.9531 |     - |     - |  12.31 KB |
|          Hero |  12.59 us | 0.028 us | 0.025 us | 0.7935 |     - |     - |   4.95 KB |

develop

|        Method |      Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |----------:|---------:|---------:|-------:|------:|------:|----------:|
| Introspection | 182.71 us | 0.312 us | 0.292 us | 3.1738 |     - |     - |  19.85 KB |
|          Hero |  12.55 us | 0.036 us | 0.034 us | 0.7935 |     - |     - |   4.95 KB |
```

**Comment by Shane32** [comment]
looks good to me

**Comment by Shane32** [comment]
only 1% of validation, but hey, every 1% counts! and about 1.5% memory reduction (of validation, for large queries).

**Comment by Shane32** [comment]
> I believe that a slight drop in performance of the most primitive queries can be allowed in favor of general normalization of work on an arbitrary query size.

I agree

**Comment by Shane32** [comment]
Well it's 4am here.  'night!

**Pull request merged**